Lib/test/test_inspect/test_inspect.py

```diff
@@ -3484,6 +3495,98 @@ def __init__(self, b):
                          ((('a', ..., ..., "positional_or_keyword"),),
                           ...))
 
+        with self.subTest('classmethod'):
+            class CM(type):
+                @classmethod
+                def __call__(cls, a):
+                    return a
+            class C(metaclass=CM):
+                def __init__(self, b):
+                    pass
+
+            self.assertEqual(C(1), 1)
+            self.assertEqual(self.signature(C),
+                            ((('a', ..., ..., "positional_or_keyword"),),
+                            ...))
+
+        with self.subTest('staticmethod'):
+            class CM(type):
+                @staticmethod
+                def __call__(a):
+                    return a
+            class C(metaclass=CM):
+                def __init__(self, b):
+                    pass
+
+            self.assertEqual(C(1), 1)
+            self.assertEqual(self.signature(C),
+                            ((('a', ..., ..., "positional_or_keyword"),),
+                            ...))
+
+        with self.subTest('MethodType'):
+            class A:
+                def call(self, a):
+                    return a
+            class CM(type):
+                __call__ = A().call
+            class C(metaclass=CM):
+                def __init__(self, b):
+                    pass
+
+            self.assertEqual(C(1), 1)
+            self.assertEqual(self.signature(C),
+                            ((('a', ..., ..., "positional_or_keyword"),),
+                            ...))
+
+        with self.subTest('partial'):
+            class CM(type):
+                __call__ = functools.partial(lambda x, a: (x, a), 2)
+            class C(metaclass=CM):
+                def __init__(self, b):
+                    pass
+
+            self.assertEqual(C(1), (2, 1))
+            self.assertEqual(self.signature(C),
+                            ((('a', ..., ..., "positional_or_keyword"),),
+                            ...))
+
+        with self.subTest('partialmethod'):
+            class CM(type):
+                __call__ = functools.partialmethod(lambda self, x, a: (x, a), 2)
+            class C(metaclass=CM):
+                def __init__(self, b):
+                    pass
+
+            self.assertEqual(C(1), (2, 1))
+            self.assertEqual(self.signature(C),
+                            ((('a', ..., ..., "positional_or_keyword"),),
+                            ...))
+
+        with self.subTest('BuiltinMethodType'):
+            class CM(type):
+                __call__ = ':'.join
+            class C(metaclass=CM):
+                def __init__(self, b):
+                    pass
+
+            self.assertEqual(C(['a', 'bc']), 'a:bc')
+            # BUG: Returns '<Signature (b)>'
```

Comment:
I have some doubts listing the bug as a test case. If we know it's a wrong answer, we should not test that the answer is accurately generated. It's possible in the future something changed and this *wrong* result is corrected, or converted to another wrong one, and the test would fail - and people will be confused unless they look at the source test code and realize it was labeled "BUG". It may help for the "backward compatibility" if we consider the bug as a compatibility we need to keep track of, but I still have concerns for this. If we do want to keep track of it, I think a `#TODO` would be much better.

Reply:
ерші is definitely a bug. The current code ignores `__call__` signature and gets the signature from `__init__`, which is not even called. I used label TODO if `signature()` currently raises ValueError, but can get a right result, and label BUG if it successfully returns wrong result. The test is added to show what the right answer is. If in the future something changed and this wrong result is corrected, it will be good, we will remove `assertRaises()` and the comment.

We can also split these tests on small methods and use the `@expectedFailure` decorator, if it matters.

BTW, I have a patch that fixes these TODOs and BUGs, but I am not certain about it, so I kept only safer changes here.